pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,253 @@
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	gcpclient "github.com/gardener/gardener-extension-provider-gcp/pkg/internal/client"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	ctrlerror "github.com/gardener/gardener/extensions/pkg/controller/error"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	gcpclient, err := a.getGCPClient(ctx, bastion)
+	if err != nil {
+		return errors.Wrap(err, "failed to create GCP client")
+	}
+
+	opt, err := DetermineOptions(ctx, bastion, cluster)
+	if err != nil {
+		return errors.Wrap(err, "failed to setup GCP options")
+	}
+
+	err = ensureFirewallRule(ctx, logger, bastion, gcpclient, opt)
+	if err != nil {
+		return errors.Wrap(err, "failed to ensure firewall rule")
+	}
+
+	endpoints, err := ensureBastionInstance(ctx, logger, gcpclient, opt)
+	if err != nil {
+		return errors.Wrap(err, "failed to ensure bastion instance")
+	}
+
+	if !endpoints.Ready() {
+		return &ctrlerror.RequeueAfterError{
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        errors.New("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	return controller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.Client(), bastion, func() error {
+		bastion.Status.Ingress = *endpoints.public
+		return nil
+	})
+
+}
+
+func ensureFirewallRule(ctx context.Context, logger logr.Logger, bastion *extensionsv1alpha1.Bastion, gcpclient gcpclient.Interface, opt *Options) error {
+	firewall, err := getFirewallRule(ctx, gcpclient, opt)
+	if err != nil {
+		return errors.Wrap(err, "could not get firewall rule")
+	}
+
+	// create firewall if it doesn't exist yet
+	if firewall == nil {
+		return createFirewallRule(ctx, bastion, gcpclient, opt)
+	}
```

Comment:
What would happen if the Firewall rule already exists, but is outdated because the CIDR(s) in the Bastion have changed?

Reply:
I may need more information or a user story about how this happened. 

in fact, I was introduced `patch` feature to update/delete the Firewall rule.  but I remove the code later as clear up handling firewall and VM instances are very complex, I haven't found a solution to handle it yet.

In a scenario, multiple users may need access to one Node at the same time (like hands-on session or live issue), they may form different Public IPs(region, home, VPN) or same Public IP (office, VPN from same region share same Public IP). I assume User A and User B both use VPN-APJ, which means they probably share one Public IP to access the node.  if User B `exit` bastion will clean up Firewall and VM, then affected UserA at the time

I suggest one bastion, one CIDR, and one Firewall rule per user.  in this way they are independent respectively.

```
apiVersion: extensions.gardener.cloud/v1alpha1
kind: Bastion
metadata:
  name: userA
...
```
```
apiVersion: extensions.gardener.cloud/v1alpha1
kind: Bastion
metadata:
  name: userB
```

testing user log `tianzhenlan-a`
```
{"level":"info","ts":"2021-07-09T09:59:43.798+0800","logger":"bastion_controller","msg":"Starting the reconciliation of bastion","bastion":"tianzhenlan-a"}
{"level":"info","ts":"2021-07-09T09:59:44.372+0800","logger":"gcp-bastion-actuator","msg":"firewall rule not found,","firewall_rule_name":"default-tianzhenlan-a-bastion-allow-ssh-access"}
{"level":"info","ts":"2021-07-09T09:59:45.106+0800","logger":"gcp-bastion-actuator","msg":"firewall created","firewall":"default-tianzhenlan-a-bastion-allow-ssh-access"}
{"level":"info","ts":"2021-07-09T09:59:45.584+0800","logger":"gcp-bastion-actuator","msg":"instance not found,","instance_name":"default-tianzhenlan-a-bastion"}
{"level":"info","ts":"2021-07-09T09:59:45.584+0800","logger":"gcp-bastion-actuator","msg":"Running new bastion instance","bastion":"default/tianzhenlan-a","operation":"reconcile"}
{"level":"info","ts":"2021-07-09T09:59:46.728+0800","logger":"gcp-bastion-actuator","msg":"disk created","bastion":"default/tianzhenlan-a","operation":"reconcile","disk":"default-tianzhenlan-a-disk"}
{"level":"info","ts":"2021-07-09T09:59:49.403+0800","logger":"bastion_controller","msg":"Successfully reconciled bastion","namespace":"default","name":"tianzhenlan-a"}
```

Reply:
> I suggest one bastion, one CIDR, and one Firewall rule per user.
not exactly, it should be per bastion instance, but this is how you implemented it, afaics.

However, @xrstf's point is valid and should be handled. The `bastion`s ingress cidr is mutable, which means the client is able to update it. 
Scenario: User wants to establish an `ssh` connection to a node and therefore creates a `bastion` resource and sets his public ip under `spec.ingress[].ipBlock.cidr` at that point in time. Now the user has to switch to the corporate network for whatever reason and connects to the respective vpn. Now he has a different ip and updates the ingress cidr on the `bastion` resource. In this case the firewall rule should be patched so that the user can continue to work with the bastion.

Reply:
committed in 207331e931675da727e95398fd32351b78dd480c